File: converted_source/usliders.pas.java
// unit Usliders

from conversion_common import *;
import delphi_compatability;

// const
kSliderHeightReadOnly = 5;
kSliderHeightNotReadOnly = 11;
kSliderDraggerHeight = 7;
kSliderDraggerWidth = 5;
kSliderGrayLineThickness = 1;
kSliderBlueLineThickness = 3;
kSliderDefaultWidth = 20;
kMinDragDistance = 1;


public void KfDrawButton(TCanvas aCanvas, TRect aRect, boolean selected, boolean enabled) {
    TRect internalRect = new TRect();
    
    aCanvas.Brush.Style = delphi_compatability.TFPBrushStyle.bsSolid;
    if (enabled) {
        aCanvas.Pen.Color = delphi_compatability.clBlack;
        aCanvas.Brush.Color = delphi_compatability.clWhite;
    } else {
        aCanvas.Brush.Color = UNRESOLVED.clBtnShadow;
        aCanvas.Pen.Color = UNRESOLVED.clBtnShadow;
    }
    aCanvas.Rectangle(aRect.Left, aRect.Top, aRect.Right, aRect.Bottom);
    if (selected) {
        internalRect = aRect;
        UNRESOLVED.inflateRect(internalRect, -1, -1);
        aCanvas.Brush.Color = delphi_compatability.clAqua;
        aCanvas.FillRect(internalRect);
    }
}


class KfSlider extends TGraphicControl {
    public TNotifyEvent FOnKeyDown;
    public TNotifyEvent FOnMouseDown;
    public TNotifyEvent FOnMouseMove;
    public TNotifyEvent FOnMouseUp;
    public long draggerHeight;
    public long draggerWidth;
    public long minValue;
    public long maxValue;
    public boolean hasUnofficialFocus;
    public boolean readOnly;
    public long currentValue;
    public boolean dragging;
    public long dragStart;
    public long originalValue;
    public TRect draggerRect;
    
    //function valueForPosition(x: longint): longint;  
    // must be at least 4 more than dragger height 
    // should be odd number 
    public void create(TComponent AnOwner) {
        super.create(AnOwner);
        this.Width = 140;
        this.Height = 35;
        this.minValue = 0;
        this.maxValue = 100;
        this.draggerHeight = kSliderDraggerHeight;
        this.draggerWidth = kSliderDraggerWidth;
        this.readOnly = false;
        this.currentValue = 0;
        this.dragging = false;
        this.useDefaultSizeAndDraggerSize();
    }
    
    public void useDefaultSizeAndDraggerSize() {
        if (this.readOnly) {
            // assumes that readOnly is already set, and assumes that value of readOnly will not be changed
            //    unless this is called again afterward 
            this.Height = kSliderHeightReadOnly;
            this.Width = kSliderDefaultWidth;
            this.draggerHeight = 0;
            this.draggerWidth = 0;
        } else {
            this.Height = kSliderHeightNotReadOnly;
            this.Width = kSliderDefaultWidth;
            this.draggerHeight = kSliderDraggerHeight;
            this.draggerWidth = kSliderDraggerWidth;
        }
    }
    
    public void CMFocusChanged(TCMFocusChanged message) {
        super.CMFocusChanged();
        //repaint;  
    }
    
    public void WMGetDlgCode(TWMGetDlgCode message) {
        message.result = delphi_compatability.DLGC_WANTARROWS;
    }
    
    //returns true if current value changed
    public boolean setValue(long newValue) {
        result = false;
        result = false;
        if (newValue > this.maxValue) {
            newValue = this.maxValue;
        }
        if (newValue < this.minValue) {
            newValue = this.minValue;
        }
        if (newValue == this.currentValue) {
            return result;
        }
        this.currentValue = newValue;
        //repaint;   
        result = true;
        return result;
    }
    
    public void doKeyDown(TObject sender, byte Key, TShiftState shift) {
        long newValue = 0;
        
        if (this.readOnly) {
            return Key;
        }
        switch (Key) {
            case delphi_compatability.VK_LEFT:
                newValue = this.currentValue - 1;
                break;
            case delphi_compatability.VK_RIGHT:
                newValue = this.currentValue + 1;
                break;
            default:
                newValue = this.currentValue;
                break;
        this.Invalidate();
        this.Refresh();
        if (this.setValue(newValue) && delphi_compatability.Assigned(this.FOnKeyDown)) {
            this.FOnKeyDown(this);
        }
        return Key;
    }
    
    public boolean pointOnDragger(long x, long y) {
        result = false;
        result = delphi_compatability.PtInRect(this.draggerRect, Point(x, y));
        return result;
    }
    
    public void mouseDown(TMouseButton Button, TShiftState Shift, int X, int Y) {
        super.mouseDown(Button, Shift, X, Y);
        if (this.readOnly) {
            if (delphi_compatability.Assigned(this.FOnMouseDown)) {
                this.FOnMouseDown(this);
            }
            return;
        }
        if ((Button == delphi_compatability.TMouseButton.mbLeft)) {
            //and (pointOnDragger(x, y))
            this.dragging = true;
            this.dragStart = X;
            this.originalValue = this.currentValue;
            //pdf fix - may want to do more here
            this.hasUnofficialFocus = true;
        }
        this.Invalidate();
        this.Refresh();
        if (delphi_compatability.Assigned(this.FOnMouseDown)) {
            this.FOnMouseDown(this);
        }
    }
    
    public void mouseMove(TShiftState Shift, int X, int Y) {
        long delta = 0;
        long newValue = 0;
        long lineLength = 0;
        
        super.mouseMove(Shift, X, Y);
        if (this.readOnly) {
            return;
        }
        if (this.dragging) {
            delta = X - this.dragStart;
            //pdf fix - won't snap to grid - will just change relative...
            lineLength = this.Width - 2 * (this.draggerWidth / 2);
            newValue = this.originalValue + intround(1.0 * delta * (this.maxValue - this.minValue) / lineLength);
            this.Invalidate();
            this.Refresh();
            if (this.setValue(newValue) && delphi_compatability.Assigned(this.FOnMouseMove)) {
                this.FOnMouseMove(this);
            }
        }
    }
    
    public void mouseUp(TMouseButton Button, TShiftState Shift, int X, int Y) {
        super.mouseUp(Button, Shift, X, Y);
        if (this.readOnly) {
            return;
        }
        if ((Button == delphi_compatability.TMouseButton.mbLeft) && this.dragging) {
            this.dragging = false;
        }
        if (abs(X - this.dragStart) < kMinDragDistance) {
            return;
        }
        this.dragStart = 0;
        if (delphi_compatability.Assigned(this.FOnMouseUp)) {
            this.FOnMouseUp(this);
        }
    }
    
    public long xDistanceFromCurrentValue(long distanceInPixels) {
        result = 0;
        float proportion = 0.0;
        
        proportion = (this.currentValue - this.minValue) / (this.maxValue - this.minValue);
        try {
            result = intround(proportion * distanceInPixels);
        } catch (Exception e) {
            result = 0;
        }
        return result;
    }
    
    public void paint() {
        TRect grayRect = new TRect();
        TRect blueRect = new TRect();
        TRect fullRect = new TRect();
        TRect lineRect = new TRect();
        long distance = 0;
        
        fullRect = this.GetClientRect();
        // start with total fill 
        this.Canvas.Brush.Color = UNRESOLVED.clBtnFace;
        this.Canvas.Brush.Style = delphi_compatability.TFPBrushStyle.bsSolid;
        this.Canvas.FillRect(fullRect);
        // rect for line 
        lineRect = fullRect;
        lineRect.Left = lineRect.Left + this.draggerWidth / 2 + 2;
        lineRect.Right = lineRect.Right - this.draggerWidth / 2 - 1;
        lineRect.Top = lineRect.Top + (lineRect.Bottom - lineRect.Top) / 2 - kSliderBlueLineThickness / 2;
        lineRect.Bottom = lineRect.Top + kSliderBlueLineThickness;
        // rect for dragger 
        distance = this.xDistanceFromCurrentValue(lineRect.Right - lineRect.Left);
        this.draggerRect.Left = lineRect.Left + distance - this.draggerWidth / 2 - this.draggerWidth % 2;
        this.draggerRect.Right = this.draggerRect.Left + this.draggerWidth;
        this.draggerRect.Top = lineRect.Top - this.draggerHeight / 2 + kSliderBlueLineThickness / 2;
        this.draggerRect.Bottom = this.draggerRect.Top + this.draggerHeight;
        // blue rect 
        blueRect = lineRect;
        blueRect.Right = blueRect.Left + distance;
        // gray rect 
        grayRect = lineRect;
        grayRect.Top = grayRect.Top + (grayRect.Bottom - grayRect.Top) / 2;
        grayRect.Bottom = grayRect.Top + kSliderGrayLineThickness;
        if (this.Enabled) {
            grayRect.Left = blueRect.Right + 1;
        }
        //draw gray line
        this.Canvas.Brush.Color = UNRESOLVED.clBtnShadow;
        this.Canvas.FillRect(grayRect);
        if (this.Enabled) {
            //draw blue line
            this.Canvas.Brush.Color = delphi_compatability.clBlue;
            this.Canvas.FillRect(blueRect);
        }
        if (!this.readOnly) {
            //draw dragger if not read only 
            KfDrawButton(this.Canvas, this.draggerRect, this.hasUnofficialFocus, this.Enabled);
        }
        if ((this.hasUnofficialFocus)) {
            this.Canvas.Pen.Color = UNRESOLVED.clBtnShadow;
            this.Canvas.MoveTo(fullRect.Right - 1, fullRect.Top);
            this.Canvas.LineTo(fullRect.Left, fullRect.Top);
            this.Canvas.LineTo(fullRect.Left, fullRect.Bottom - 1);
            this.Canvas.LineTo(fullRect.Right - 1, fullRect.Bottom - 1);
            this.Canvas.LineTo(fullRect.Right - 1, fullRect.Top);
        }
    }
    
}

File: converted_source/usliders.pas.py
# unit Usliders

from conversion_common import *
import delphi_compatability

# const
kSliderHeightReadOnly = 5
kSliderHeightNotReadOnly = 11
kSliderDraggerHeight = 7
kSliderDraggerWidth = 5
kSliderGrayLineThickness = 1
kSliderBlueLineThickness = 3
kSliderDefaultWidth = 20
kMinDragDistance = 1

def KfDrawButton(aCanvas, aRect, selected, enabled):
    internalRect = TRect()
    
    aCanvas.Brush.Style = delphi_compatability.TFPBrushStyle.bsSolid
    if enabled:
        aCanvas.Pen.Color = delphi_compatability.clBlack
        aCanvas.Brush.Color = delphi_compatability.clWhite
    else:
        aCanvas.Brush.Color = UNRESOLVED.clBtnShadow
        aCanvas.Pen.Color = UNRESOLVED.clBtnShadow
    aCanvas.Rectangle(aRect.Left, aRect.Top, aRect.Right, aRect.Bottom)
    if selected:
        internalRect = aRect
        UNRESOLVED.inflateRect(internalRect, -1, -1)
        aCanvas.Brush.Color = delphi_compatability.clAqua
        aCanvas.FillRect(internalRect)

class KfSlider(TGraphicControl):
    def __init__(self):
        self.FOnKeyDown = TNotifyEvent()
        self.FOnMouseDown = TNotifyEvent()
        self.FOnMouseMove = TNotifyEvent()
        self.FOnMouseUp = TNotifyEvent()
        self.draggerHeight = 0L
        self.draggerWidth = 0L
        self.minValue = 0L
        self.maxValue = 0L
        self.hasUnofficialFocus = false
        self.readOnly = false
        self.currentValue = 0L
        self.dragging = false
        self.dragStart = 0L
        self.originalValue = 0L
        self.draggerRect = TRect()
    
    #function valueForPosition(x: longint): longint;  
    # must be at least 4 more than dragger height 
    # should be odd number 
    def create(self, AnOwner):
        TGraphicControl.create(self, AnOwner)
        self.Width = 140
        self.Height = 35
        self.minValue = 0
        self.maxValue = 100
        self.draggerHeight = kSliderDraggerHeight
        self.draggerWidth = kSliderDraggerWidth
        self.readOnly = false
        self.currentValue = 0
        self.dragging = false
        self.useDefaultSizeAndDraggerSize()
        return self
    
    def useDefaultSizeAndDraggerSize(self):
        if self.readOnly:
            # assumes that readOnly is already set, and assumes that value of readOnly will not be changed
            #    unless this is called again afterward 
            self.Height = kSliderHeightReadOnly
            self.Width = kSliderDefaultWidth
            self.draggerHeight = 0
            self.draggerWidth = 0
        else:
            self.Height = kSliderHeightNotReadOnly
            self.Width = kSliderDefaultWidth
            self.draggerHeight = kSliderDraggerHeight
            self.draggerWidth = kSliderDraggerWidth
    
    def CMFocusChanged(self, message):
        TGraphicControl.CMFocusChanged(self)
        #repaint;  
    
    def WMGetDlgCode(self, message):
        message.result = delphi_compatability.DLGC_WANTARROWS
    
    #returns true if current value changed
    def setValue(self, newValue):
        result = false
        result = false
        if newValue > self.maxValue:
            newValue = self.maxValue
        if newValue < self.minValue:
            newValue = self.minValue
        if newValue == self.currentValue:
            return result
        self.currentValue = newValue
        #repaint;   
        result = true
        return result
    
    def doKeyDown(self, sender, Key, shift):
        newValue = 0L
        
        if self.readOnly:
            return Key
        if Key == delphi_compatability.VK_LEFT:
            newValue = self.currentValue - 1
        elif Key == delphi_compatability.VK_RIGHT:
            newValue = self.currentValue + 1
        else :
            newValue = self.currentValue
        self.Invalidate()
        self.Refresh()
        if self.setValue(newValue) and delphi_compatability.Assigned(self.FOnKeyDown):
            self.FOnKeyDown(self)
        return Key
    
    def pointOnDragger(self, x, y):
        result = false
        result = delphi_compatability.PtInRect(self.draggerRect, Point(x, y))
        return result
    
    def mouseDown(self, Button, Shift, X, Y):
        TGraphicControl.mouseDown(self, Button, Shift, X, Y)
        if self.readOnly:
            if delphi_compatability.Assigned(self.FOnMouseDown):
                self.FOnMouseDown(self)
            return
        if (Button == delphi_compatability.TMouseButton.mbLeft):
            #and (pointOnDragger(x, y))
            self.dragging = true
            self.dragStart = X
            self.originalValue = self.currentValue
            #pdf fix - may want to do more here
            self.hasUnofficialFocus = true
        self.Invalidate()
        self.Refresh()
        if delphi_compatability.Assigned(self.FOnMouseDown):
            self.FOnMouseDown(self)
    
    def mouseMove(self, Shift, X, Y):
        delta = 0L
        newValue = 0L
        lineLength = 0L
        
        TGraphicControl.mouseMove(self, Shift, X, Y)
        if self.readOnly:
            return
        if self.dragging:
            delta = X - self.dragStart
            #pdf fix - won't snap to grid - will just change relative...
            lineLength = self.Width - 2 * (self.draggerWidth / 2)
            newValue = self.originalValue + intround(1.0 * delta * (self.maxValue - self.minValue) / lineLength)
            self.Invalidate()
            self.Refresh()
            if self.setValue(newValue) and delphi_compatability.Assigned(self.FOnMouseMove):
                self.FOnMouseMove(self)
    
    def mouseUp(self, Button, Shift, X, Y):
        TGraphicControl.mouseUp(self, Button, Shift, X, Y)
        if self.readOnly:
            return
        if (Button == delphi_compatability.TMouseButton.mbLeft) and self.dragging:
            self.dragging = false
        if abs(X - self.dragStart) < kMinDragDistance:
            return
        self.dragStart = 0
        if delphi_compatability.Assigned(self.FOnMouseUp):
            self.FOnMouseUp(self)
    
    def xDistanceFromCurrentValue(self, distanceInPixels):
        result = 0L
        proportion = 0.0
        
        proportion = (self.currentValue - self.minValue) / (self.maxValue - self.minValue)
        try:
            result = intround(proportion * distanceInPixels)
        except:
            result = 0
        return result
    
    def paint(self):
        grayRect = TRect()
        blueRect = TRect()
        fullRect = TRect()
        lineRect = TRect()
        distance = 0L
        
        fullRect = self.GetClientRect()
        # start with total fill 
        self.Canvas.Brush.Color = UNRESOLVED.clBtnFace
        self.Canvas.Brush.Style = delphi_compatability.TFPBrushStyle.bsSolid
        self.Canvas.FillRect(fullRect)
        # rect for line 
        lineRect = fullRect
        lineRect.Left = lineRect.Left + self.draggerWidth / 2 + 2
        lineRect.Right = lineRect.Right - self.draggerWidth / 2 - 1
        lineRect.Top = lineRect.Top + (lineRect.Bottom - lineRect.Top) / 2 - kSliderBlueLineThickness / 2
        lineRect.Bottom = lineRect.Top + kSliderBlueLineThickness
        # rect for dragger 
        distance = self.xDistanceFromCurrentValue(lineRect.Right - lineRect.Left)
        self.draggerRect.Left = lineRect.Left + distance - self.draggerWidth / 2 - self.draggerWidth % 2
        self.draggerRect.Right = self.draggerRect.Left + self.draggerWidth
        self.draggerRect.Top = lineRect.Top - self.draggerHeight / 2 + kSliderBlueLineThickness / 2
        self.draggerRect.Bottom = self.draggerRect.Top + self.draggerHeight
        # blue rect 
        blueRect = lineRect
        blueRect.Right = blueRect.Left + distance
        # gray rect 
        grayRect = lineRect
        grayRect.Top = grayRect.Top + (grayRect.Bottom - grayRect.Top) / 2
        grayRect.Bottom = grayRect.Top + kSliderGrayLineThickness
        if self.Enabled:
            grayRect.Left = blueRect.Right + 1
        #draw gray line
        self.Canvas.Brush.Color = UNRESOLVED.clBtnShadow
        self.Canvas.FillRect(grayRect)
        if self.Enabled:
            #draw blue line
            self.Canvas.Brush.Color = delphi_compatability.clBlue
            self.Canvas.FillRect(blueRect)
        if not self.readOnly:
            #draw dragger if not read only 
            KfDrawButton(self.Canvas, self.draggerRect, self.hasUnofficialFocus, self.Enabled)
        if (self.hasUnofficialFocus):
            self.Canvas.Pen.Color = UNRESOLVED.clBtnShadow
            self.Canvas.MoveTo(fullRect.Right - 1, fullRect.Top)
            self.Canvas.LineTo(fullRect.Left, fullRect.Top)
            self.Canvas.LineTo(fullRect.Left, fullRect.Bottom - 1)
            self.Canvas.LineTo(fullRect.Right - 1, fullRect.Bottom - 1)
            self.Canvas.LineTo(fullRect.Right - 1, fullRect.Top)
    
